### ISSUE TITLE
Some Notes

### ISSUE THREAD

**Issue by alexander-g** [abrarum/bezierobjdet#1]
Hi,

I've taken a look at your commits, but I have difficulties to follow your thought process or what you intend to do. Could I ask you to add a bit more documentation for every commit? This can be either directly at the beginning in the `.ipynb` files, one or two bullet points what you wanted to achieve with this code. Alternatively you can add commit coments here on github.

Next, I suggest you should try to work with the small line dataset that you've already created in `index.ipynb`. For starters it's perfectly fine, that it the lines only have two control points. My suggestion is to use a very simple network: a backbone (mobilenet/resnet without the last classification layer) and an additional layer that simply predicts/regresses the two control points. This might be already not trivial because the order of the points matters.

Alternatively, if you have a plan of your own, feel free to share it.

**Comment by abrarum** [comment]
I'm so sorry about that. Because of my experimentation, I created multiple folders but I've ordered them now. I'll remove the unnecessary folders and comment the files. Btw, the route I'm working on is: "tests/tf_begin/" which is the root, here you can find the scripts.py which is the file I'm working on and has all the functions to pre-process and train etc, the output and the models etc can be found in: "tests/tf_begin/workspace/**SDL**". I'll briefly write you later about what's the flow and what I'm achieving. continued....

EDIT: name "**SDL**" is changed to "resnet".

**Comment by abrarum** [comment]
I've now ordered the folders. TESTS is now on the root having 3 custom script .ipynb files with obvious names. Inside "workspace/resnet" is the folder where all the datasets, annotations, pre-trained models and exported-models are. "./MODELS/my_resenet" is the folder where the checkpoints are but since they are more than 100mb and github has restrictions, I have ignored them.

IDEA:
Initially I started with MASKRCNN and then switched to FasterRCNN as you suggested. There is a problem in running fcnn from the repo and its still in discussion to resolve it, I then quickly switched to simple resnet (./pre-trained-models), prepared tfrecords ("./annotations") from images and XML and trained the model ("exported-modelsmy_model").

> that it the lines only have two control points

In reality the bezier formula needs 4 points (for straight lines) just as needed in drawing a box: xmin, xmax, ymin, ymax but I'll use these 4 points as a feature trained and generate a curve (straight line) out of this. Here is an example of bezier functions I'm using to generate a curve outta 4 points:

![nodes_preview](https://user-images.githubusercontent.com/7055212/112146519-f3d73c80-8bdb-11eb-92b5-3fda5f08176a.png)

Secondly, you suggested mobileresnet, I'm already using a resnet, can check it out the mobile too if needed. I then used the pre-trained weights from it as you're suggesting. I've some results regarding the training but it isn't good. The training started with a loss of 4000 and finished with 40k steps with a loss of 20. I haven't evaluated the model yet on my test images but I'll let you know..

**EDIT**:

1. I've added the comments on the 3 scripts.

**Comment by alexander-g** [comment]
Now this might sound like a stupid question, but: Why are you doing this? Why do you train a normal SSD object detector on your lines dataset?

I see two reasons why you might be doing this:
- Maybe you want to establish a baseline for future evaluation with you own algorithm. This is fine, but then you should not spend too much time on it and move on to the "own" part sometime soon.
- Maybe you misunderstand the topic a bit. The core idea of this thesis is not to simply use an existing box-based object detector and train it on a new dataset, this is not something new and not enough for a master's thesis. The core idea is rather to replace the anchor boxes with anchor (bezier) lines and see how this works out. For this, I'm not sure how well suited the TensorFlow Object Detection API is. I'm afraid it might be based too much on boxes, though I have not looked too much into it yet.

If you want we could also set up a video meeting and discuss directly, maybe (hopefully) I just misunderstand your thought process.

**Comment by abrarum** [comment]
Hi. I just didn't want to start with my own code and wanted to test it on something that is already available and I've seen how worst its performance and cluttering the code has become. Your second point is exactly what I'm going for right now but I think it's better as you suggested to be on the same page. When are you free for a call?, I can then make myself available.

**Comment by alexander-g** [comment]
I had something like this in mind regarding the line regression model:
![line_regressor](https://user-images.githubusercontent.com/3867427/112637866-e2419f00-8e3e-11eb-8cb9-bc2caf28b107.png)

In TensorFlow/Keras this would be simply:
```python
backbone         = keras.applications.MobileNet(include_top=False, pooling='avg')
regression_layer = keras.layers.Dense(4)

x = x0   = keras.Input(shape=(None,None,3))
x        = backbone(x)
x        = regression_layer(x)
model    = keras.Model(inputs=x0, outputs=x)
```

Edit: it should x0,y0,x1,y1 in the image
